near-contract-standards/src/access/mod.rs

```diff
@@ -1,129 +1,279 @@
+    ///
+    /// Has no effect if the [`role`](Self::roles) does not exist,
+    /// or if the [`account`](RoleData::members) is not a member for that `role`,
+    /// or if the `account` was already disabled.
+    fn internal_revoke_role(&mut self, role: RoleId, account: AccountId) {
+        use std::collections::hash_map::Entry;
+        match self.roles.entry(role) {
+            // `role` exists
+            Entry::Occupied(mut role_entry) => {
+                match role_entry.get_mut().members.entry(account) {
+                    // `account` is a member, update status
+                    Entry::Occupied(mut member_entry) => {
+                        *member_entry.get_mut() = false;
+                    }
+
+                    // `account` isn't a member, no action needed
+                    Entry::Vacant(_) => {}
+                }
+            }
+
+            // `role` inexistent, no action needed
+            Entry::Vacant(_) => {}
         }
     }
 
-    pub fn renounce_role(&mut self, role: [u8; 32], account: AccountId) {
+    /// Revokes [`role` from](RoleData::members) the [predecessor's `account`](env::predecessor_account_id()).  
+    /// Has no effect if the [`role`](Self::roles) does not exist, or if the [`account`
+    /// is not enabled](RoleData::members) for that `role`.
+    ///
+    /// This method's purpose is to provide a mechanism for `account`s to purposefuly
+    /// [lose their own privileges](RoleData::members), such as when they are compromised
+    /// (eg. when a trusted device is misplaced).
+    ///
+    /// Requirements:
+    ///
+    /// - The `account` must be the [predecessor account](env::predecessor_account_id()).
+    ///
+    /// See also [`Self::grant_role()`] and [`Self::revoke_role()`] for other management of
+    /// roles.
+    pub fn renounce_role(&mut self, role: RoleId, account: AccountId) {
         require!(
             account == env::predecessor_account_id(),
             "AccessControl: can only renounce roles for self"
         );
-        self.revoke_role(role, account);
+        self.internal_revoke_role(role, account);
     }
 
-    pub fn set_role_admin(&mut self, role: [u8; 32], admin_role: [u8; 32]) {
-        if !self.roles.contains_key(&role) {
-            self.roles.insert(
-                role,
-                RoleData { members: HashMap::new(), admin_role: self.default_admin_role.clone() },
-            );
+    /// Sets `admin_role` as `role`'s [admin role](RoleData::admin_role),
+    /// [creating the `role`](Self::roles) if necessary.  
+    ///
+    /// # Warning
+    ///
+    /// There are no further verifications about the caller.
+    fn internal_set_role_admin(&mut self, role: RoleId, admin_role: RoleId) {
+        use std::collections::hash_map::Entry;
+        match self.roles.entry(role) {
+            // `role` does not exist
+            Entry::Vacant(role_entry) => {
+                // creates the role with the correct admin
```

Comment:
This is an effect change as previously it didn't create roles. But I'm not sure which way it should be